File: app/src/main/java/com/test/mobile_test/base/BaseViewModel.java
package com.test.mobile_test.base;

import android.content.Context;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.test.mobile_test.data.DataManager;

import io.reactivex.disposables.CompositeDisposable;

public class BaseViewModel<T> extends ViewModel {
    protected Context context;
    protected MutableLiveData<T> mutableLiveData;
    protected CompositeDisposable compositeDisposable = new CompositeDisposable();
    protected DataManager mDataManager;

    public BaseViewModel(Context context, DataManager mDataManager) {
        this.context = context;
        this.mDataManager = mDataManager;
        mutableLiveData = new MutableLiveData<>();
    }

    public MutableLiveData<T> getMutableLiveData() {
        return mutableLiveData;
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        if (compositeDisposable != null) {
            compositeDisposable.dispose();
        }
    }
}

File: app/src/main/java/com/ztd/interview_test/mvvm/base/BaseViewModel.kt
package com.ztd.interview_test.mvvm.base

import androidx.lifecycle.ViewModel
import io.reactivex.disposables.CompositeDisposable
import java.lang.ref.WeakReference

/**
 * Created by Mahdi_ZareTahghighDoost(ZTD)
 *  on 1/10/2020.
 */

abstract class BaseViewModel<N> : ViewModel() {

    var compositeDisposable: CompositeDisposable = CompositeDisposable()
    lateinit var mNavigator: WeakReference<N>

    fun setNavigator(navigator: N) {
        this.mNavigator = WeakReference(navigator)
    }

    override fun onCleared() {
        compositeDisposable.dispose()
        super.onCleared()
    }

}